Issue 593 Improve activeWhen to support query-strings in URL

Sauloxd opened the pull request:
# What does this PR do?
This implements the fix in issue #593 
In order to support both query-strings in pathname and in hash pathname, I opted for cleaning the route before trying to match with the regex. Now we always remove location.search and EVERYTHING after the '?' I'm assuming, after the location.search is removed, the remaining query-string is from hash pathname, so it's safe to just remove everything after it.

Also, refactor a bit the tests on `pathToActiveWhen` to make it easier to read them, since I added a bunch of other tests, and having Jest asserts around would make it impossible to reason about each `it`. Also changed the test to actually use pathToActiveWhen, and by doing so, I found out another edge case:
- When the app is declared on a activeWhen like `activeWhen: 'pathname'` currently the implementation does not match for `http://app.com/pathname` and will only work if you use `activeWhen: '/pathname'`. Should we improve the regex generator to cover for this scenario?
![image](https://user-images.githubusercontent.com/11878589/86082447-23eb1800-ba6e-11ea-8e79-f0aa13739dcf.png)

Sauloxd commented on spec/apis/path-to-active-when.spec.js:
```diff
@@ -2,182 +2,121 @@ import { pathToActiveWhen } from "single-spa";
+    });
+
+    // Impossible paths due to URL always goin from root 'app.com' -> 'app.com/'
+    expectPathToMatch("#/subpath/:dynamic", {
```
As far as I searched, we the location will always have the `/` prefix, so when we declare an activeWhen with `#/subpath` the URL will automatically becomes `app.com/#/subpath` (only if it's root). I'm not sure what to do, I'm really open to some help here. Is it possible to the following asserts be true?

joeldenning replied:
This?

```js
if (activeWhen.startsWith('#')) {
  activeWhen = '/' + activeWhen
}
```

Sauloxd replied:
Yeah, I'm not sure how to handle those cases, was waiting for some feedback on what to do!

When testing just the regex generator in isolation '#/subpath/:dynamic' really generates a regex able to match strings like '#/subpath/1
But when testing the actual pathToActiveWhen (that needs a URL as argument) this test would never pass (even without my changes) making me wonder if paths without starting with `/` ever worked in real scenarios (based on how `location` object works):

Just to be extra clear, I'm thinking about this scenario:
I declared an App with activeWhen: `'#/app1'`
In the browser, when I go to `https://bleus.com/#/app1`, the App would never mount (because the regex generated by `#/app1` wouldn't match `https://bleus.com/#/app1`)

When developing, `new Url('https://bleus.com#/app1')` would give me a location.href == 'https://bleus.com/#/app1', automatically adding a slash between `com` and `#`. So when we do:
``` js
location.href
      .replace(location.origin, "")
```
it would never (as far as I know at least) give me `#/app1` but always somethings starting with `/`...

joeldenning replied:
I think it might be best to automatically prepend a `/` when we see an `activeWhen` that starts with `#`. That way our users don't have to worry about this nuanced edge case.

joeldenning commented on spec/apis/path-to-active-when.spec.js:
```diff
@@ -2,182 +2,121 @@ import { pathToActiveWhen } from "single-spa";
+    });
+
+    expectPathToMatch("pathname", {
+      "http://app.com/pathname": false,
```
Why is this false? Because it's missing the leading slash?

Sauloxd replied:
Yes, when the appToActiveWhen manipulates the `location.href`, there will always give a route starting with `/`, so unless we improve these cases where the path doesn't start with '/', in more real life scenarios it wouldn't pass

joeldenning commented on spec/apis/path-to-active-when.spec.js:
```diff
@@ -2,182 +2,121 @@ import { pathToActiveWhen } from "single-spa";
+    });
+
+    expectPathToMatch(":dynamic/:dynamic", {
+      "http://app.com/1/1": false,
```
Why is this false? In the [previous tests](https://github.com/single-spa/single-spa/pull/594/files#diff-a769be383da96d054df67ad869d0f04dL134), it was true

Sauloxd replied:
The previous test was testing a different part of the code, just the regex generator... If `:dynamic/:dynamic` started with a `/`, this test would be true

joeldenning replied:
I see

joeldenning commented on spec/apis/path-to-active-when.spec.js:
```diff
@@ -2,182 +2,121 @@ import { pathToActiveWhen } from "single-spa";
-      );
-    });
-
-    it("Should generate correct regex for '$🎉/:dynamic$🎉'", () => {
```
Was this test intentionally deleted?

Sauloxd replied:
Yes, this is debatable, although it's possible I thought having only real world scenarios here would make the tests better... I could return them, no problem 👍

joeldenning replied:
My understanding is the web URLs only support ascii characters and don't support emojis, so I think this is fine to omit.

Sauloxd commented:
Updated PR with your suggestions:
- Instead of `startsWith`, opted for a simple [0] for IE11 compatibility
- Updated tests: Now activeWhen that don't start with `/` passes like expected
